src/messages/MClientReply.h

```diff
@@ -134,69 +141,116 @@ struct InodeStat {
     decode(p, features);
   }
 
-  void decode(bufferlist::const_iterator &p, uint64_t features) {
+  void decode(bufferlist::const_iterator &p, uint64_t features = 0) {
     using ceph::decode;
-    decode(vino.ino, p);
-    decode(vino.snapid, p);
-    decode(rdev, p);
-    decode(version, p);
-    decode(xattr_version, p);
-    decode(cap, p);
-    {
-      ceph_file_layout legacy_layout;
-      decode(legacy_layout, p);
-      layout.from_legacy(legacy_layout);
-    }
-    decode(ctime, p);
-    decode(mtime, p);
-    decode(atime, p);
-    decode(time_warp_seq, p);
-    decode(size, p);
-    decode(max_size, p);
-    decode(truncate_size, p);
-    decode(truncate_seq, p);
-    decode(mode, p);
-    decode(uid, p);
-    decode(gid, p);
-    decode(nlink, p);
-    decode(dirstat.nfiles, p);
-    decode(dirstat.nsubdirs, p);
-    decode(rstat.rbytes, p);
-    decode(rstat.rfiles, p);
-    decode(rstat.rsubdirs, p);
-    decode(rstat.rctime, p);
-
-    decode(dirfragtree, p);
-
-    decode(symlink, p);
-    
-    if (features & CEPH_FEATURE_DIRLAYOUTHASH)
+    uint64_t pre;
+    decode(pre, p);
+    if (pre == (uint64_t)-1) {
```

Comment:
You can add an argument to this method to note whether the connection supports versioned inode stats:

`void decode(bufferlist::const_iterator &p, uint64_t features = 0, bool versioned = false)`

Then the callers can check the session features to see if the MDS supports `CEPHFS_FEATURE_INODEDIRSTAT` and pass that as the bool.

Reply:
@batrick 
We are encoding `-1` to let the client know (InodeStat::decode), if the MDS supports the `CEPHFS_FEATURE_REPLY_ENCODING`(the new name). The additional bool argument would be unnecessary, as this happens without it now. As I understand, we have a plan to drop the `features` later too. 

So could you please check again and clarify? 

Thanks.

Reply:
patrick means let caller of InodeStat/DirStat/LeaseDir::decode to decide if new format is used. For example:

void InodeStat::decode(bufferlist::const_iterator &p, uint64_t features) {
  if (features == (uint64_t)-1) {
    DECODE_START()
    ...
    DECODE_END();
  } else {
    decode(ino, p)
    ...
  }
}

Client.cc code should be:

if (session->mds_features.test(CEPHFS_FEATURE_ REPLY_ENCODING)
   features = (uint64_t)-1;
....
InodeStat ist(p, features););

Reply:
@ukernel @batrick Fixed.